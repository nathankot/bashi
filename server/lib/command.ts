export * from "./command/types.ts";
export * from "./command/parser.ts";
export * from "./command/builtinCommands.ts";
export * from "./command/filter.ts";
export * from "./command/run.ts";
export * from "./command/supportedModels.ts";
