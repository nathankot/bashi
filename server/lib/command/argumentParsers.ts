import * as t from "io-ts";
import { parseDate } from "chrono";

import { StringValue } from "@lib/valueTypes.ts";

export type ArgumentParserContext = {
  now: Date;
  chronoParseDate: typeof parseDate;
  timezoneUtcOffset: number;
};

export const argumentParsers = {
  naturalLanguageDateTime: {
    inputType: "string" as const,
    outputType: "string" as const,
    fn: function (ctx: ArgumentParserContext, arg: string): null | StringValue {
      const d = ctx.chronoParseDate(arg, {
        instant: ctx.now,
        timezone: ctx.timezoneUtcOffset,
      });
      if (d == null) {
        return null;
      }
      return {
        type: "string",
        value: d.toISOString(),
      };
    },
  },
};

export const ArgumentParser = t.keyof(argumentParsers);
export type ArgumentParser = t.TypeOf<typeof ArgumentParser>;

export default argumentParsers;
