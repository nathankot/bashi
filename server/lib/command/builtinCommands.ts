import * as mathjs from "mathjs";
import { parseDate } from "chrono";
import { formatInTimeZone, getTimezoneOffset } from "date-fns-tz";

import { PROGRAMMING_LANGUAGES } from "@lib/constants.ts";
import { BuiltinCommandDefinition } from "./types.ts";
import { HTTPError } from "@lib/errors.ts";
import { run as runPassthrough } from "@lib/models/passthroughOpenai000.ts";
import { run as runCode } from "@lib/models/code000.ts";
import { run as runTranslate } from "@lib/models/translate000.ts";

const LOCAL_DATETIME_FORMAT = "yyyy-MM-dd'T'HH:mm:ssXXX";

const now: BuiltinCommandDefinition<[], "string"> = {
  isBuiltin: true,
  cost: -1000,
  description: "get current ISO8601 datetime",
  args: [],
  run: async (deps, []) => ({
    type: "string",
    value: formatInTimeZone(
      deps.now(),
      deps.session.configuration.timezoneName,
      LOCAL_DATETIME_FORMAT
    ),
  }),
  returnType: "string",
};

const parseRelativeTime: BuiltinCommandDefinition<["string"], "string"> = {
  isBuiltin: true,
  cost: 10,
  returnType: "string",
  description: "get ISO8601 datetime relative to now from natural language",
  args: [
    {
      name: "natural language relative time",
      type: "string",
    },
  ],
  run: async (deps, [description]) => {
    const d = parseDate(description.value, {
      instant: deps.now(),
      // chronojs expects the offset in minutes:
      timezone:
        getTimezoneOffset(deps.session.configuration.timezoneName) / 1000 / 60,
    });
    if (d == null) {
      throw new HTTPError("could not parse relative time", 500);
    }
    return {
      type: "string",
      value: formatInTimeZone(
        d,
        deps.session.configuration.timezoneName,
        LOCAL_DATETIME_FORMAT
      ),
    };
  },
};

const currentTimeForTimezone: BuiltinCommandDefinition<["string"], "string"> = {
  isBuiltin: true,
  cost: -1000,
  returnType: "string",
  description: `get ISO8601 datetime for the given timezone`,
  args: [{ name: "tz database timezone", type: "string" }],
  triggerTokens: ["time", "hour", "clock"],
  run: async ({ log, session, now }, [timeZone]) => {
    return {
      type: "string",
      value: formatInTimeZone(
        now(),
        timeZone.value.replaceAll(" ", "_"),
        LOCAL_DATETIME_FORMAT
      ),
    };
  },
};

const topWebsitesForQuery: BuiltinCommandDefinition<["string"], "string"> = {
  isBuiltin: true,
  cost: 1000,
  description: "get website titles and links from the search query",
  returnType: "string",
  args: [{ name: "query", type: "string" }],
  run: async (modelDeps, [query]) => {
    const results = await modelDeps.googleSearch(query.value, modelDeps.signal);
    return {
      type: "string",
      value: results
        .map((r) => `Title: ${r.title}\nLink: ${r.link}\nSnippet: ${r.snippet}`)
        .join("\n---\n"),
    };
  },
  triggerTokens: ["search", "google", "find"],
};

const extractInformation: BuiltinCommandDefinition<
  ["string", "string"],
  "string"
> = {
  isBuiltin: true,
  cost: 100,
  returnType: "string",
  description: "describe/summarize/extract information from the given string",
  args: [
    {
      name: "full description of desired output and its format",
      type: "string",
    },
    { name: "input text/code", type: "string" },
  ],
  run: async (modelDeps, [desc, input]) => {
    const output = await runPassthrough(
      modelDeps,
      { model: "passthrough-openai-000" },
      {
        openAiModel: "text-davinci-003",
        request: `Condense/summarize/extract information from text that follows. The output should satisfy the requirement:\n${desc.value}\n\nThe text is:\n${input.value}\n\nResult:\n"`,
      }
    );
    return {
      type: "string",
      value: output.result,
    };
  },
};

const math: BuiltinCommandDefinition<["string"], "string"> = {
  isBuiltin: true,
  cost: -1000,
  returnType: "string",
  description: `get the result of a math formula`,
  args: [{ name: "formula (ascii characters only)", type: "string" }],
  run: async ({ log, session }, [expr]) => {
    try {
      const result = mathjs.evaluate(expr.value);
      return { type: "string", value: `${result satisfies string}` };
    } catch (e) {
      log("info", "failed to parse math: " + expr.value);
      throw new HTTPError(e, 500);
    }
  },
};

const translate: BuiltinCommandDefinition<["string", "string"], "string"> = {
  isBuiltin: true,
  cost: 100,
  returnType: "string",
  description: `translate text into target language`,
  args: [
    { name: "input", type: "string" },
    { name: "target language full name", type: "string" },
  ],
  run: async (modelDeps, [input, targetLanguage]) => {
    const output = await runTranslate(
      modelDeps,
      { model: "translate-000" },
      {
        request: input.value,
        targetLanguage: targetLanguage.value,
      }
    );
    return { type: "string", value: output.result.trim() };
  },
  triggerTokens: [
    "translate",
    "say",
    "how",
    "write",
    "language",
    "japanese",
    "french",
    "italian",
    "spanish",
    "chinese",
    "english",
    "arabic",
    "portuguese",
    "creole",
    "dutch",
    "danish",
    "catalan",
    "german",
    "greek",
    "icelandic",
    "norwegian",
    "swedish",
    "polish",
    "russian",
    "ukrainian",
    "hungarian",
    // TODO use complete list
  ],
};

const editText: BuiltinCommandDefinition<["string", "string"], "string"> = {
  isBuiltin: true,
  cost: 100,
  returnType: "string",
  description: `edit text using the given requirements. prefer editCode() for code`,
  args: [
    { name: "text to edit", type: "string" },
    {
      name: "full description of desired changes or additions",
      type: "string",
    },
  ],
  run: async (modelDeps, [text, editingRequirement]) => {
    const output = await runPassthrough(
      modelDeps,
      { model: "passthrough-openai-000" },
      {
        openAiModel: "text-davinci-003",
        request: `Rewrite and edit the following text. The requirement is '${
          editingRequirement.value satisfies string
        }':
${text.value}

The re-written text is:
`,
      }
    );
    return { type: "string", value: output.result.trim() };
  },
  triggerTokens: [
    "edit",
    "change",
    "alter",
    "fix",
    "move",
    "align",
    "reword",
    "re-word",
    "editor",
    "improve",
    "check",
    "revise",
    "modify",
    "adapt",
    "rewrite",
    "re-write",
  ],
};

// TODO: maybe a code edit model will do better with this?
const editCode: BuiltinCommandDefinition<
  ["string", "string", "string"],
  "string"
> = {
  isBuiltin: true,
  cost: 100,
  returnType: "string",
  description: `edit code using the given requirements`,
  args: [
    { name: "code to edit", type: "string" },
    { name: "programming language or 'unknown'", type: "string" },
    {
      name: "full description of desired changes or additions",
      type: "string",
    },
  ],
  run: async (modelDeps, [text, language, editingRequirement]) => {
    const output = await runPassthrough(
      modelDeps,
      { model: "passthrough-openai-000" },
      {
        openAiModel: "text-davinci-003",
        request: `Edit or refactor the code below based on the given requirement.
Programming language is '${language.value satisfies string}'.
The requirement is '${editingRequirement.value satisfies string}':
${text.value satisfies string}

The full edited code is:
`,
      }
    );
    return { type: "string", value: output.result.trim() };
  },
  triggerTokens: [
    ...PROGRAMMING_LANGUAGES,
    "function",
    "compile",
    "class",
    "variable",
    "code",
    "move",
    "align",
    "refactor",
    "re-factor",
    "edit",
    "change",
    "alter",
    "fix",
    "reword",
    "re-word",
    "editor",
    "lint",
    "improve",
    "check",
    "revise",
    "modify",
    "adapt",
    "rewrite",
    "re-write",
  ],
};

const writeCommitMessage: BuiltinCommandDefinition<["string"], "string"> = {
  isBuiltin: true,
  cost: 1000,
  returnType: "string",
  description: `generate a git commit title/message based on diff from input text`,
  args: [{ name: "diff", type: "string" }],
  run: async (modelDeps, [diff]) => {
    const output = await runPassthrough(
      modelDeps,
      { model: "passthrough-openai-000" },
      {
        openAiModel: "text-davinci-003",
        request: `Write a short (under 50 characters) git commit title as best you can by summarizing the following diff:

\`\`\`diff
${diff.value}
\`\`\`

Result:
`,
      }
    );
    return { type: "string", value: output.result.trim() };
  },
  triggerTokens: ["commit", "git", "diff"],
};

const generateCode: BuiltinCommandDefinition<["string", "string"], "string"> = {
  isBuiltin: true,
  cost: 10,
  returnType: "string",
  description: `generate code for the given language and verbose description/requirements`,
  args: [
    { name: "language", type: "string" },
    {
      name: "requirements",
      type: "string",
    },
  ],
  run: async (modelDeps, [targetLanguage, request]) => {
    const output = await runCode(
      modelDeps,
      { model: "code-000" },
      {
        request: request.value,
        programmingLanguage: targetLanguage.value,
      }
    );
    return { type: "string", value: output.result.trim() };
  },
  triggerTokens: [
    ...PROGRAMMING_LANGUAGES,
    "generate",
    "code",
    "programming",
    "program",
    "write",
    "swift",
    "html",
    "lang",
    "function",
    "class",
    "language",
    "type",
  ],
};

export const builtinCommands = {
  now,
  math,
  currentTimeForTimezone,
  editText,
  editCode,
  generateCode,
  writeCommitMessage,
  translate,
  parseRelativeTime,
  extractInformation,
  topWebsitesForQuery,
};

export default builtinCommands;
