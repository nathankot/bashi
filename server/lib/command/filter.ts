import * as p from "typescript-parsec";

import {
  buildLexer,
  rule,
  expectEOF,
  expectSingleResult,
} from "typescript-parsec";

import { CommandSet } from "./types.ts";

enum TokenKind {
  Word,
  Space,
  StopWord,
  Other,
}

const stopWords = {
  i: null,
  me: null,
  my: null,
  myself: null,
  we: null,
  our: null,
  ours: null,
  ourselves: null,
  you: null,
  your: null,
  yours: null,
  yourself: null,
  yourselves: null,
  he: null,
  him: null,
  his: null,
  himself: null,
  she: null,
  her: null,
  hers: null,
  herself: null,
  it: null,
  its: null,
  itself: null,
  they: null,
  them: null,
  their: null,
  theirs: null,
  themselves: null,
  what: null,
  which: null,
  who: null,
  whom: null,
  this: null,
  that: null,
  these: null,
  those: null,
  am: null,
  is: null,
  are: null,
  was: null,
  were: null,
  be: null,
  been: null,
  being: null,
  have: null,
  has: null,
  had: null,
  having: null,
  do: null,
  does: null,
  did: null,
  doing: null,
  a: null,
  an: null,
  the: null,
  and: null,
  but: null,
  if: null,
  or: null,
  because: null,
  as: null,
  until: null,
  while: null,
  of: null,
  at: null,
  by: null,
  for: null,
  with: null,
  about: null,
  against: null,
  between: null,
  into: null,
  through: null,
  during: null,
  before: null,
  after: null,
  above: null,
  below: null,
  to: null,
  from: null,
  up: null,
  down: null,
  in: null,
  out: null,
  on: null,
  off: null,
  over: null,
  under: null,
  again: null,
  further: null,
  then: null,
  once: null,
  here: null,
  there: null,
  when: null,
  where: null,
  why: null,
  how: null,
  all: null,
  any: null,
  both: null,
  each: null,
  few: null,
  more: null,
  most: null,
  other: null,
  some: null,
  such: null,
  no: null,
  nor: null,
  not: null,
  only: null,
  own: null,
  same: null,
  so: null,
  than: null,
  too: null,
  very: null,
  s: null,
  t: null,
  can: null,
  will: null,
  just: null,
  don: null,
  should: null,
  now: null,
};

const lexer = buildLexer([
  [
    false,
    new RegExp(`^(${Object.keys(stopWords).join("|")})`, "g"),
    TokenKind.StopWord,
  ],
  [true, /^[A-z'-]*/g, TokenKind.Word],
  [false, /^\s+/g, TokenKind.Space],
  [false, /^[^A-z'-]*/g, TokenKind.Other],
]);

const WORDS = rule<TokenKind, Record<string, true>>();
WORDS.setPattern(
  p.apply(p.rep_sc(p.tok(TokenKind.Word)), (toks) => {
    let result: Record<string, true> = {};
    for (const tok of toks) {
      result[tok.text.toLowerCase()] = true;
    }
    return result;
  })
);

export function parseWordsFromRequest(request: string): Record<string, true> {
  return expectSingleResult(expectEOF(WORDS.parse(lexer.parse(request))));
}

export function filterUnnecessary(
  request: string,
  commandSet: CommandSet
): CommandSet {
  const result: CommandSet = {};
  const words = parseWordsFromRequest(request);
  commandLoop: for (const [k, c] of Object.entries(commandSet)) {
    if ("disable" in c && c.disable === true) {
      continue commandLoop;
    }
    if (c.triggerTokens == null || c.triggerTokens.length === 0) {
      result[k] = c;
      continue commandLoop;
    }
    for (const trigger of c.triggerTokens) {
      if (words[trigger.toLowerCase()] === true) {
        result[k] = c;
        continue commandLoop;
      }
    }
  }
  return result;
}
