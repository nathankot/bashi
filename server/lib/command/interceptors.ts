import math from "./interceptors/math.ts";
import time from "./interceptors/time.ts";
import translate from "./interceptors/translate.ts";
import generateCode from "./interceptors/generateCode.ts";
import editProse from "./interceptors/editProse.ts";
import editCode from "./interceptors/editCode.ts";

export const commandInterceptors = {
  math,
  time,
  translate,
  generateCode,
  editProse,
  editCode,
};
