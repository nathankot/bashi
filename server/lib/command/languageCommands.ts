import { ValueType, ValueForType } from "@lib/valueTypes.ts";
import {
  AnyBuiltinCommandDefinition,
  BuiltinCommandDefinition,
} from "./types.ts";

export const languageCommands = {
  $ref: {
    isBuiltin: true,
    description: "retrieve a variable",
    args: [{ name: "identifier", type: "string" }],
    returnType: "mixed",
    run: async (_, [iden], memory) => {
      const maybeValue = memory.variables[iden.value];
      if (maybeValue != null) {
        return maybeValue;
      }
      if (iden.value.toLowerCase() === "result") {
        const maybeLastResult = [...memory.topLevelResults]
          .reverse()
          .filter((r) => r.type !== "void")[0];
        if (maybeLastResult != null) {
          return maybeLastResult;
        }
      }
      throw new Error(`the variable '${iden.value}' does not exist`);
    },
  } as BuiltinCommandDefinition<["string"], "mixed">,
  "__=__": {
    overloads: ValueType.types
      .map((t) => t.value)
      .map(
        (t): AnyBuiltinCommandDefinition => ({
          isBuiltin: true,
          description: "assign a variable",
          args: [
            { name: "identifier", type: "string" },
            { name: "value", type: t },
          ],
          returnType: t,
          run: async (_, [lhs, rhs], memory) => {
            memory.variables[lhs.value] = rhs;
            return rhs;
          },
        })
      ),
  },
  "__+__": {
    overloads: [
      overload("number", "number", "number", (l, r) => ({
        type: "number",
        value: l.value + r.value,
      })),
      overload("string", "string", "string", (l, r) => ({
        type: "string",
        value: l.value + r.value,
      })),
      overload("number", "string", "string", (l, r) => ({
        type: "string",
        value: l.value.toString() + r.value,
      })),
      overload("string", "number", "string", (l, r) => ({
        type: "string",
        value: l.value + r.value.toString(),
      })),
      overload("boolean", "string", "string", (l, r) => ({
        type: "string",
        value: (l.value === true ? "true" : "false") + r.value,
      })),
      overload("string", "boolean", "string", (l, r) => ({
        type: "string",
        value: l.value + (r.value === true ? "true" : "false"),
      })),
    ],
  },
};

export function overload<
  LHS extends ValueType,
  RHS extends ValueType,
  R extends ValueType
>(
  lhs: LHS,
  rhs: RHS,
  returnType: R,
  fn: (lhs: ValueForType<LHS>, rhs: ValueForType<RHS>) => ValueForType<R>
): AnyBuiltinCommandDefinition {
  return {
    isBuiltin: true,
    description: `overload for ${lhs}, ${rhs} => ${returnType}`,
    args: [
      { name: "lhs", type: lhs },
      { name: "rhs", type: rhs },
    ],
    returnType,
    run: async (_, [lhs, rhs]) => fn(lhs, rhs),
  };
}
