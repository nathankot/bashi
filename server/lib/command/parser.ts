import * as t from "io-ts";
import * as p from "typescript-parsec";

import {
  // Token,
  buildLexer,
  rule,
  expectEOF,
  expectSingleResult,
} from "typescript-parsec";

import { Value } from "@lib/valueTypes.ts";

interface Call {
  type: "call";
  name: string;
  args: (Value | Call)[];
}

const Call: t.Type<Call> = t.recursion("Call", () =>
  t.type({
    type: t.literal("call"),
    name: t.string,
    args: t.array(t.union([Value, Call])),
  })
);

export const ActionGroup = t.intersection([
  t.type({
    thought: t.string,
    action: t.string,
    functionCalls: t.array(Call),
  }),
  t.partial({
    result: t.string,
  }),
]);
export type ActionGroup = t.TypeOf<typeof ActionGroup>;

enum ActionTokenKind {
  KeywordThought,
  KeywordAction,
  KeywordResult,

  String,

  Newline,
  Space,
}

const actionLexer = buildLexer([
  [true, /^Action/gi, ActionTokenKind.KeywordAction],
  [true, /^Thought/gi, ActionTokenKind.KeywordThought],
  [true, /^Result/gi, ActionTokenKind.KeywordResult],

  [true, /^:.*/g, ActionTokenKind.String],

  [true, /^(\n|\r\n|\r)+/g, ActionTokenKind.Newline],
  [false, /^\s+/g, ActionTokenKind.Space],
]);

const ACTION_GROUP = rule<ActionTokenKind, ActionGroup>();
ACTION_GROUP.setPattern(
  p.apply(
    p.seq(
      p.apply(
        p.kright(
          p.tok(ActionTokenKind.KeywordThought),
          p.tok(ActionTokenKind.String)
        ),
        (str) => str.text.slice(1).trim()
      ),
      p.tok(ActionTokenKind.Newline),
      p.apply(
        p.kright(
          p.tok(ActionTokenKind.KeywordAction),
          p.tok(ActionTokenKind.String)
        ),
        (str) => str.text.slice(1).trim()
      ),
      p.opt_sc(
        p.apply(
          p.kright(
            p.tok(ActionTokenKind.Newline),
            p.kright(
              p.tok(ActionTokenKind.KeywordResult),
              p.tok(ActionTokenKind.String)
            )
          ),
          (str) => {
            const result = str.text.slice(1).trim();
            if (result.length === 0) {
              return undefined;
            }
            return result;
          }
        )
      )
    ),
    ([thought, , action, result]) => {
      const functionCalls = parseFunctionCalls(action);
      return {
        thought,
        action,
        result,
        functionCalls,
      };
    }
  )
);

export function parseActionGroup(expr: string): ActionGroup {
  return expectSingleResult(
    expectEOF(ACTION_GROUP.parse(actionLexer.parse(expr)))
  );
}

enum FunctionTokenKind {
  Identifier,

  TrueLiteral,
  FalseLiteral,
  NumberLiteral,
  SingleQuoteStringLiteral,
  DoubleQuoteStringLiteral,
  BackQuoteStringLiteral,

  LParen,
  RParen,
  Space,
  Comma,
  SemiColon,
}

const functionLexer = buildLexer([
  [true, /^'([^'\\]|\\.)*'/g, FunctionTokenKind.SingleQuoteStringLiteral],
  [true, /^"([^"\\]|\\.)*"/g, FunctionTokenKind.DoubleQuoteStringLiteral],
  [true, /^`([^`\\]|\\.)*`/g, FunctionTokenKind.BackQuoteStringLiteral],
  [true, /^[\+\-]?\d+(\.\d+)?/g, FunctionTokenKind.NumberLiteral],
  [true, /^true/g, FunctionTokenKind.TrueLiteral],
  [true, /^false/g, FunctionTokenKind.FalseLiteral],

  [true, /^[a-zA-Z_-][a-zA-Z0-9_-]*/g, FunctionTokenKind.Identifier],

  [true, /^\(/g, FunctionTokenKind.LParen],
  [true, /^\)/g, FunctionTokenKind.RParen],
  [true, /^\,/g, FunctionTokenKind.Comma],
  [true, /^;/g, FunctionTokenKind.SemiColon],
  [false, /^\s+/g, FunctionTokenKind.Space],
]);

const VALUE = rule<FunctionTokenKind, Value>();
VALUE.setPattern(
  p.apply(
    p.alt(
      p.tok(FunctionTokenKind.TrueLiteral),
      p.tok(FunctionTokenKind.FalseLiteral),
      p.tok(FunctionTokenKind.NumberLiteral),
      p.tok(FunctionTokenKind.SingleQuoteStringLiteral),
      p.tok(FunctionTokenKind.DoubleQuoteStringLiteral),
      p.tok(FunctionTokenKind.BackQuoteStringLiteral)
    ),
    (tok) => {
      switch (tok.kind) {
        case FunctionTokenKind.TrueLiteral:
          return { type: "boolean", value: true };
        case FunctionTokenKind.FalseLiteral:
          return { type: "boolean", value: false };
        case FunctionTokenKind.NumberLiteral:
          return { type: "number", value: +tok.text };
        case FunctionTokenKind.SingleQuoteStringLiteral:
          return {
            type: "string",
            value: tok.text
              .substring(1, tok.text.length - 1)
              .replaceAll(`\\'`, `'`),
          };
        case FunctionTokenKind.DoubleQuoteStringLiteral:
          return {
            type: "string",
            value: tok.text
              .substring(1, tok.text.length - 1)
              .replaceAll(`\\"`, `"`),
          };
        case FunctionTokenKind.BackQuoteStringLiteral:
          return {
            type: "string",
            value: tok.text
              .substring(1, tok.text.length - 1)
              .replaceAll("\\`", "`"),
          };
        default:
          const exhaustiveCheck: never = tok;
          throw new Error(
            `Unsupported value kind: ${(exhaustiveCheck as any).text}`
          );
      }
    }
  )
);

const CALL = rule<FunctionTokenKind, Call>();
CALL.setPattern(
  p.apply(
    p.seq(
      p.tok(FunctionTokenKind.Identifier),
      p.tok(FunctionTokenKind.LParen),
      p.opt_sc(p.list(p.alt(VALUE, CALL), p.tok(FunctionTokenKind.Comma))),
      p.tok(FunctionTokenKind.RParen)
    ),
    ([{ text: name }, , maybeArgs]) => ({
      type: "call",
      name,
      args: maybeArgs ?? [],
    })
  )
);

const CALLS = rule<FunctionTokenKind, Call[]>();
CALLS.setPattern(
  p.apply(
    p.kleft(
      p.list(
        CALL,
        p.seq(
          p.tok(FunctionTokenKind.SemiColon),
          p.rep_sc(p.tok(FunctionTokenKind.SemiColon))
        )
      ),
      p.rep_sc(p.tok(FunctionTokenKind.SemiColon))
    ),
    (calls) => calls
  )
);

export function parseFunctionCall(expr: string): Call {
  return expectSingleResult(expectEOF(CALL.parse(functionLexer.parse(expr))));
}

export function parseFunctionCalls(expr: string): Call[] {
  return expectSingleResult(expectEOF(CALLS.parse(functionLexer.parse(expr))));
}
