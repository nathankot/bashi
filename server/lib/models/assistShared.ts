import * as t from "io-ts";

import { Command, CommandExecuted, CommandParsed, Expr } from "@lib/command.ts";
import { Value } from "@lib/valueTypes.ts";

export const Input = t.partial({
  request: t.string,
  resolvedCommands: t.record(t.string, Value),
});
export type Input = t.TypeOf<typeof Input>;

export const ResultFinished = t.type({
  type: t.literal("finished"),
  results: t.array(Value),
});
export type ResultFinished = t.TypeOf<typeof ResultFinished>;

export const ResultPendingCommands = t.type({
  type: t.literal("pending_commands"),
  pendingCommands: t.array(Command),
  results: t.array(Value),
});
export type ResultPendingCommands = t.TypeOf<typeof ResultPendingCommands>;

export const Result = t.union([ResultFinished, ResultPendingCommands]);
export type Result = t.TypeOf<typeof Result>;

export function getPendingCommandsOrResult(
  commandId: string,
  expr: Expr,
  resolvedCommmands: Record<string, CommandExecuted>
): { pendingCommands: CommandParsed[] } | { result: Value } {
  // If we have an terminal value:
  if (expr.type !== "call") {
    return { result: expr };
  }

  // See if its already resolved
  const maybeAlreadyResolved = resolvedCommmands[commandId];
  if (maybeAlreadyResolved != null) {
    if (maybeAlreadyResolved.name != expr.name) {
      throw new Error(
        `corruption! expected resolved call to be ${expr.name} ` +
          `but got ${maybeAlreadyResolved.name}`
      );
    }
    const result = maybeAlreadyResolved.returnValue;
    return {
      result,
    };
  }

  // Build up a list of nested commands that must be resolved,
  // or if they are all resolved build up a list of resolved arguments.
  let pendingCommands: CommandParsed[] = [];
  let resolvedArguments: Value[] | null = [];
  for (const [i, arg] of Object.entries(expr.args)) {
    if (arg.type !== "call") {
      resolvedArguments?.push(arg);
      continue;
    }
    const argResult = getPendingCommandsOrResult(
      commandId + "." + i,
      arg,
      resolvedCommmands
    );
    if ("pendingCommands" in argResult) {
      pendingCommands = [...pendingCommands, ...argResult.pendingCommands];
      // if at least 1 arg is still pending, then we don't have resolved arguments:
      resolvedArguments = null;
      continue;
    }
    resolvedArguments?.push(argResult.result);
  }

  // If all of the arguments are resolved then this call
  // becomes pending:
  if (resolvedArguments != null) {
    if (resolvedArguments.length !== expr.args.length) {
      throw new Error(
        `corruption! for ${expr.name} expected ${expr.args.length} ` +
          `arguments but got ${resolvedArguments.length}`
      );
    }
    return {
      pendingCommands: [
        {
          type: "parsed",
          args: resolvedArguments,
          id: commandId,
          name: expr.name,
        },
      ],
    };
  }

  // Otherwise we need to resolve the nested children first:
  return { pendingCommands };
}
