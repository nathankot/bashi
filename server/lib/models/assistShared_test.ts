import { assertSnapshot } from "std/testing/snapshot.ts";

import { CommandExecuted, parseExpression } from "@lib/command.ts";
import { getPendingCommandsOrResult } from "./assistShared.ts";

for (const test of [
  {
    description: "example A step 1",
    commandId: "0",
    call: parseExpression(`test(a(), b(123, c()))`),
    resolvedCommands: {},
  },
  {
    description: "example A step 2",
    commandId: "0",
    call: parseExpression(`test(a(), b(123, c()))`),
    resolvedCommands: {
      "0.0": {
        type: "executed",
        args: [],
        id: "0.0",
        name: "a",
        returnValue: { type: "number", value: 123 },
      },
      "0.1.1": {
        type: "executed",
        args: [],
        id: "0.1.1",
        name: "c",
        returnValue: { type: "string", value: "blah" },
      },
    } as Record<string, CommandExecuted>,
  },
  {
    description: "example A step 3",
    commandId: "0",
    call: parseExpression(`test(a(), b(123, c()))`),
    resolvedCommands: {
      "0.0": {
        type: "executed",
        args: [],
        id: "0.0",
        name: "a",
        returnValue: { type: "number", value: 123 },
      },
      "0.1": {
        type: "executed",
        args: [],
        id: "0.1",
        name: "b",
        returnValue: { type: "string", value: "ha" },
      },
    } as Record<string, CommandExecuted>,
  },
  {
    description: "example A step 4",
    commandId: "0",
    call: parseExpression(`test(a(), b(123, c()))`),
    resolvedCommands: {
      "0": {
        type: "executed",
        args: [],
        id: "0",
        name: "test",
        returnValue: { type: "number", value: 123 },
      },
    } as Record<string, CommandExecuted>,
  },
]) {
  Deno.test(test.description, async (t) => {
    try {
      await assertSnapshot(
        t,
        getPendingCommandsOrResult(
          test.commandId,
          test.call,
          test.resolvedCommands
        )
      );
    } catch (e) {
      await assertSnapshot(t, e.message);
    }
  });
}
