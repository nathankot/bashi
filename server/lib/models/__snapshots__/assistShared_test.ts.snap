export const snapshot = {};

snapshot[`example A step 1 1`] = `
{
  pendingCommands: [
    {
      args: [
      ],
      id: "0.0",
      name: "a",
      type: "parsed",
    },
    {
      args: [
      ],
      id: "0.1.1",
      name: "c",
      type: "parsed",
    },
  ],
}
`;

snapshot[`example A step 2 1`] = `
{
  pendingCommands: [
    {
      args: [
        {
          type: "number",
          value: 123,
        },
        {
          type: "string",
          value: "blah",
        },
      ],
      id: "0.1",
      name: "b",
      type: "parsed",
    },
  ],
}
`;

snapshot[`example A step 3 1`] = `
{
  pendingCommands: [
    {
      args: [
        {
          type: "number",
          value: 123,
        },
        {
          type: "string",
          value: "ha",
        },
      ],
      id: "0",
      name: "test",
      type: "parsed",
    },
  ],
}
`;

snapshot[`example A step 4 1`] = `
{
  result: {
    type: "number",
    value: 123,
  },
}
`;
